File: backend/src/application/usecase/DeleterProductUseCase.go
package usecase

import (
	"fmt"

	"github.com/marcosfrancomarinho/go-first-project/src/application/dto"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/interfaces"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/valuesobject"
)

type DeleterProductUseCase struct {
	deleterProduct interfaces.DeleterProduct
}

func NewDeleterProductUseCase(deleterProduct interfaces.DeleterProduct) *DeleterProductUseCase {
	return &DeleterProductUseCase{deleterProduct}
}

func (d *DeleterProductUseCase) Delete(payload *dto.RequestDeleterProductDTO) (*dto.ResponseDeleterProductDTO, error) {
	id, err := valuesobject.NewID(payload.Id)
	if err != nil {
		return nil, err
	}
	if err := d.deleterProduct.Delete(id); err != nil {
		return nil, err
	}
	return &dto.ResponseDeleterProductDTO{
		Id:      id.GetValue(),
		Message: fmt.Sprintf("product Id: %s deletado com sucesso", id.GetValue()),
	}, nil
}

File: frontend/src/application/usecase/DeleterProductUseCase.ts
import type { HttpDeleteClient } from '../../domain/gateway/HttpDeleteClient';
import type { StorageClient } from '../../domain/gateway/StorageClient';
import { ID } from '../../domain/valuesobject/ID';
import { TokenError } from '../../shared/erros/TokenError';
import type { RequestDeleterProductDTO } from '../dto/RequestDeleterProductDTO';
import type { ResponseDeleterProductDTO } from '../dto/ResponseDeleterProductDTO';

export class DeleterProductUseCase {
  public constructor(
    private httpClient: HttpDeleteClient,
    private storageClient: StorageClient,
    private path: string,
    private key: string
  ) {}

  public async delete(payload: RequestDeleterProductDTO): Promise<ResponseDeleterProductDTO> {
    try {
      const id: ID = ID.create(payload.id);
      const { token } = this.storageClient.get<{ token: string }>(this.key);
      const exclusionResponse: ResponseDeleterProductDTO = await this.httpClient.delete(this.path, id.getValue(), { token });
      return exclusionResponse;
    } catch (error) {
      error instanceof TokenError && this.storageClient.delete(this.key);
      throw error;
    }
  }
}
